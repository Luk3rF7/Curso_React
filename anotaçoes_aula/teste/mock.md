<h1 align="center"> Mock service  work</h1>

<h2>Utilizameremos biblioteca:</h2>

<p style="font-size:16px" >
Mock service work uma da mais utilizada pelo professor luiz miranda, <a href="https://mswjs.io/docs/getting-started/install"> link </a>
<p>

<li style="font-size:16px"> Comando para install: npm i msw --save-dev 

<h3> Teste assincrono </h3>
<hr>
<p style="font-size:16px"> 
  Quando temos teste assincronos geralmente precisamo utilizar : 


        describe('<Home />', () => {
          it('should render search,posts and load more',async () => {
            const { debug } = render(<Home />);
            const noMorePosts = screen.getByText('não existe posts');
            await waitForElementToBeRemoved(noMorePosts);
            debug();
             });
          });


 <li style="font-size:16px"> utilizamos o waitForElementToBeRemoved(passado onde tem requisiçao da api);

 a gente cria os handles,passarei para ele quais os tipo que url
 com isso nos fazemos nossa importação de 2 importante da 
 libs msw :
</li>
<li  style="font-size:18px">
 rest from msw & setupServer from msw/node 
</li>
<hr>

            import { render, screen, waitForElementToBeRemoved } from '@testing-library/react';
            import { Home } from './Home';
            import { rest } from 'msw';
            import { setupServer } from 'msw/node';

            const handles = [];
            const server = setupServer(...handles);
            describe('<Home />', () => {
              it('should render search,posts and load more', async () => {
                render(<Home />);
                const noMorePosts = screen.getByText('não existe posts');
                await waitForElementToBeRemoved(noMorePosts);
                screen.debug();
              });
            });

<li style="font-size:17px"> os handles são para cada chamada url exemplo :  </li>

    const handles = [
      rest.get('url do post', async(req,res,ctx) => {
        return res(ctx.json(
          adiciono os post ,seleciono no maximo 5 p/ teste
        ));
      })
    ];

<li  style="font-size:15px"> servidor de teste </li>

Para iniciar servidor utilizo onde ele vai escutar :
<hr>

                    beforeAll(() => {
                      server.listen();
                      });
Para não atrapalhar certa interrupção ou teste afete outro utilizo :
<hr>

                    afterEach(() => {
                      server.resetHandlers()
                      });

Para encerrar o servidor depois de tudo utilizo :
<hr>

                    afterAll(
                      () => {
                        server.close();
                        });


<li style="font-size:16px"> Podemos utilizar no url '*https://jsonplaceholder.typicode.com/*' assim ele ira buscar todos da url: </li>
</p>
