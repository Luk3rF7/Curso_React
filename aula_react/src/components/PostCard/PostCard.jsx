import './PostCard.css';
import Props from 'prop-types';

export const PostCard = ({ 
  title,
  cover,
  body,
  id
}) => {
    return (
         <div className="post">
          <img src={cover} alt={title} />
          <div 
            className="post-content">
                <h1>{title}</h1>
           <p> {body} </p> 
             </div> 
        </div> 
    )
}
PostCard.propTypes = {
  title: Props.string.isRequired,
  cover: Props.string.isRequired,
  body: Props.string.isRequired,
  id: Props.number.isRequired,
}
