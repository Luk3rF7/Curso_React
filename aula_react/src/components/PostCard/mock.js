export const postCardPropMock = {
  title: 'title_1 ',
  body: 'body 1',
  id: 1,
  cover: 'img/img.png'
};