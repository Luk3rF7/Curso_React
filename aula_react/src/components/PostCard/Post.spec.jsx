import { PostCard } from './PostCard';
import { postCardPropMock } from './mock';
const { render,screen  } = require('@testing-library/react');


const props = postCardPropMock;
describe('<PostCard />', () => {
  it('should render PostCard correctly', () => {
   render(<PostCard {...props} />);
   /* img */
   expect(screen.getByRole('img' , { name: /title_1/i}))
   .toHaveAttribute('src','img/img.png');
    /* heading */
   expect(screen.getByRole('heading' , { name: 'title_1'}))
   .toBeInTheDocument();
   /*  texto */
   expect(screen.getByText('body 1'))
   .toBeInTheDocument();
  });
  it('should match snapshot', () => {
    const {container} = render(<PostCard {...props} />);
    expect(container.firstChild).toMatchSnapshot();

  })
})