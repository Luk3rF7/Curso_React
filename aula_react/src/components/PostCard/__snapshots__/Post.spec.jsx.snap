// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostCard /> should match snapshot 1`] = `
<div
  class="post"
>
  <img
    alt="title_1 "
    src="img/img.png"
  />
  <div
    class="post-content"
  >
    <h1>
      title_1 
    </h1>
    <p>
       
      body 1
       
    </p>
  </div>
</div>
`;
